crates/ra_project_model/src/lib.rs

```diff
@@ -406,18 +407,49 @@ fn find_rust_project_json(path: &Path) -> Option<PathBuf> {
     None
 }
 
-fn find_cargo_toml(path: &Path) -> Result<PathBuf> {
-    if path.ends_with("Cargo.toml") {
-        return Ok(path.to_path_buf());
-    }
+fn find_cargo_toml_down_the_fs(path: &Path) -> Option<PathBuf> {
     let mut curr = Some(path);
     while let Some(path) = curr {
         let candidate = path.join("Cargo.toml");
         if candidate.exists() {
-            return Ok(candidate);
+            return Some(candidate);
         }
         curr = path.parent();
     }
+    
+    None
+}
+
+fn find_cargo_toml_up_the_fs(path: &Path) -> Option<PathBuf> {
+    let entities = match read_dir(path) {
+        Ok(entities) => entities,
+        Err(_) => return None
+    };
+
+    // Only one level up to avoid cycles the easy way and stop a runaway scan with large projects
+    for entity in entities.filter_map(Result::ok) {
+        let candidate = entity.path().join("Cargo.toml");
+        if candidate.exists() {
+            return Some(candidate);
+        }
+    }
+
+    None
+}
+
+fn find_cargo_toml(path: &Path) -> Result<PathBuf> {
+    if path.ends_with("Cargo.toml") {
+        return Ok(path.to_path_buf());
+    }
+
+    if let Some(p) = find_cargo_toml_down_the_fs(path) {
+        return Ok(p)
+    }
+
+    if let Some(p) = find_cargo_toml_up_the_fs(path) {
+        return Ok(p)
+    }
+
     Err(CargoTomlNotFoundError(path.to_path_buf()).into())
 }
 
```

Comment:
Yeah, the naming is confusing... Let's do `find_in_parent_dir`, `find_in_child_dir` mabye?

Reply:
I think you might have meant this anyway but just mentioning that I prefixed with find_cargo_toml_... because these functions are in a fairly "large" file so a new reader might wonder - find what?